Ensure terminal pods maintain terminal status

ehashman opened the pull request:
Alternative to #105411. This is a slightly tighter patch and appears to also fix the issue.

#### What type of PR is this?

/kind bug
/sig node
/priority critical-urgent

#### What this PR does / why we need it:

Avoids overwriting statuses or phase on terminated pods.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #105358

#### Special notes for your reviewer 

This avoids an `isTerminated` check, which we may not need. We can fix the container restart count issue separately.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Evicted and other terminated pods will no longer revert to Running phase
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ehashman commented:
/test pull-kubernetes-node-kubelet-eviction
/test pull-kubernetes-node-kubelet-serial

ehashman committed:
Do not sync Waiting statuses for Terminated pods

ehashman committed:
Ensure terminal pods maintain terminal status

ehashman commented on pkg/kubelet/status/status_manager_test.go:
```diff
@@ -622,6 +622,64 @@ func TestTerminatePod(t *testing.T) {
 	assert.Equal(t, newStatus.Message, firstStatus.Message)
 }
 
+func TestTerminatePodWaiting(t *testing.T) {
```
Basically copied from test above, but changed (init)container 3 to waiting status

smarterclayton commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -2595,6 +2595,42 @@ func Test_generateAPIPodStatus(t *testing.T) {
 				},
 			},
 		},
+		{
```
Add another condition that swaps pod and previousStatus, and then one that has restart never and all containers have currentStatus = terminated

ehashman replied:
The first is already covered with https://github.com/kubernetes/kubernetes/blob/c32e56912852a2f902426394fdbe5950176e1eae/pkg/kubelet/kubelet_pods_test.go#L2563-L2597 but we don't have anything for restart never so adding that now

ehashman replied:
Pushed one more unit test to cover the other case.

rphillips commented:
/triage accepted

ehashman committed:
Fix bugs in e2e pod test

- Update test to account for late synced statuses
- Terminated containers are not running
- Add missing format values to test logline

rphillips commented:
/retest-required

ehashman commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -2595,6 +2595,85 @@ func Test_generateAPIPodStatus(t *testing.T) {
+					{Type: v1.PodScheduled, Status: v1.ConditionTrue},
+				},
+				ContainerStatuses: []v1.ContainerStatus{
+					ready(succeededState("containerA")),
+					ready(succeededState("containerB")),
```
I don't actually think these should be ready but neither should any of the waiting containers in the previous existing cases, I think this test just wasn't written correctly. See also containers from pods that are failed showing restarts.

matthyx replied:
Should we fix these tests first, and then apply your patch instead of having 1) ineffective coverage or 2) changing fixing kubelet and tests in the same PR?

smarterclayton replied:
Yeah, the "isTerminated" change might be what I expect to squash that (a terminated container status should never be reported as ready, except for init containers that reached completion).  Can track it in the restart one.

ehashman replied:
> Should we fix these tests first, and then apply your patch instead of having 1) ineffective coverage or 2) changing fixing kubelet and tests in the same PR?

Given the urgency of this fix, and how much difficulty I had with the unit tests yesterday, I don't think we should wait. I am more confident in the e2es than the unit tests for exercising this logic.

smarterclayton commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -2595,6 +2595,85 @@ func Test_generateAPIPodStatus(t *testing.T) {
 				},
 			},
 		},
+		{
```
I expected both of these to test that message and reason are preserved from previous state

ehashman replied:
Done!

ehashman commented:
/test pull-kubernetes-verify-govet-levee

ehashman commented:
flake https://github.com/kubernetes/kubernetes/issues/99877
/test pull-kubernetes-unit

ehashman commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/105462/pull-kubernetes-e2e-kind/1445425277617311744
> /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/node/pods.go:205
Oct  5 16:51:17.002: 1 errors:
pod pod-submit-status-2-13 on node kind-worker2 container unexpected exit code 137: start=0001-01-01 00:00:00 +0000 UTC end=0001-01-01 00:00:00 +0000 UTC reason=ContainerStatusUnknown message=The container could not be located when the pod was terminated
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113

I think this e2e needs to be updated, pods being force killed should be fine (but it wasn't written to be tolerant of those statuses getting synced out).

smarterclayton commented on test/e2e/node/pods.go:
```diff
@@ -379,12 +379,14 @@ var _ = SIGDescribe("Pods Extended", func() {
 								switch {
 								case t.ExitCode == 1:
 									// expected
+								case t.ExitCode == 137 && (t.Reason == "ContainerStatusUnknown" || t.Reason == "Error"):
```
Did this start happening with your PR?

smarterclayton replied:
If so, was it because previously we would have left the container in waiting state during graceful shutdown?

ehashman replied:
> Did this start happening with your PR?

Yes.

> If so, was it because previously we would have left the container in waiting state during graceful shutdown?

Either waiting or running I think. Statuses weren't properly getting reported.

smarterclayton commented:
Concerning failure:

pod pod-submit-status-0-13 on node kind-worker did not reach a terminal phase before being deleted but had running containers: phase=Pending, running-containers=busybox

Happening on other tests?

ehashman commented:
I see some instances of failures in the periodics but they don't appear to match: https://storage.googleapis.com/k8s-triage/index.html?test=Pods%20Extended

ehashman commented:
Unrelated flake:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

ehashman commented:
```
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.668493     248 config.go:383] "Receiving a new pod" pod="pods-7829/pod-submit-status-0-13"
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.669167     248 kubelet.go:2055] "SyncLoop ADD" source="api" pods=[pods-7829/pod-submit-status-0-13]
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.669398     248 pod_workers.go:535] "Pod is being synced for the first time" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.669503     248 pod_workers.go:742] "Processing pod event" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7 updateType=0
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.669541     248 kubelet.go:1533] "syncPod enter" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.669559     248 kubelet_pods.go:1437] "Generating pod status" pod="pods-7829/pod-submit-status-0-13"
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.669600     248 kubelet_pods.go:1449] "Got phase for pod" pod="pods-7829/pod-submit-status-0-13" oldPhase=Pending phase=Pending
Oct 05 17:41:03 kind-worker kubelet[248]: I1005 17:41:03.694537     248 kubelet.go:2071] "SyncLoop DELETE" source="api" pods=[pods-7829/pod-submit-status-0-13]
```

The node got a DELETE before it even started to spin up the pod... looks like it wasn't able to sync out the final status due to client-side throttling:

```
Oct 05 17:41:05 kind-worker kubelet[248]: I1005 17:41:05.527801     248 pod_workers.go:931] "Pod is complete and the worker can now stop" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7
Oct 05 17:41:05 kind-worker kubelet[248]: I1005 17:41:05.527832     248 pod_workers.go:811] "Processing pod event done" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7 updateType=2
Oct 05 17:41:09 kind-worker kubelet[248]: I1005 17:41:09.645158     248 request.go:597] Waited for 386.835047ms due to client-side throttling, not priority and fairness, request: GET:https://kind-control-plane:6443/api/v1/namespaces/pods-7829/pods/pod-submit-status-0-13
Oct 05 17:41:09 kind-worker kubelet[248]: I1005 17:41:09.843151     248 request.go:597] Waited for 154.773669ms due to client-side throttling, not priority and fairness, request: PATCH:https://kind-control-plane:6443/api/v1/namespaces/pods-7829/pods/pod-submit-status-0-13/status
Oct 05 17:41:09 kind-worker kubelet[248]: I1005 17:41:09.860368     248 status_manager.go:622] "Patch status for pod" pod="pods-7829/pod-submit-status-0-13" patch="{\"metadata\":{\"uid\":\"a409f0f6-ac57-4d5e-95e1-97e074ea74b7\"},\"status\":{\"$setElementOrder/conditions\":[{\"type\":\"Initialized\"},{\"type\":\"Ready\"},{\"type\":\"ContainersReady\"},{\"type\":\"PodScheduled\"}],\"conditions\":[{\"lastProbeTime\":null,\"lastTransitionTime\":\"2021-10-05T17:41:03Z\",\"status\":\"True\",\"type\":\"Initialized\"},{\"lastProbeTime\":null,\"lastTransitionTime\":\"2021-10-05T17:41:03Z\",\"message\":\"containers with unready status: [busybox]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"Ready\"},{\"lastProbeTime\":null,\"lastTransitionTime\":\"2021-10-05T17:41:03Z\",\"message\":\"containers with unready status: [busybox]\",\"reason\":\"ContainersNotReady\",\"status\":\"False\",\"type\":\"ContainersReady\"}],\"containerStatuses\":[{\"image\":\"k8s.gcr.io/e2e-test-images/busybox:1.29-2\",\"imageID\":\"\",\"lastState\":{},\"name\":\"busybox\",\"ready\":false,\"restartCount\":0,\"started\":false,\"state\":{\"terminated\":{\"exitCode\":137,\"finishedAt\":null,\"message\":\"The container could not be located when the pod was terminated\",\"reason\":\"ContainerStatusUnknown\",\"startedAt\":null}}}],\"hostIP\":\"172.18.0.2\",\"podIP\":\"10.244.1.111\",\"podIPs\":[{\"ip\":\"10.244.1.111\"}],\"startTime\":\"2021-10-05T17:41:03Z\"}}"
Oct 05 17:41:09 kind-worker kubelet[248]: I1005 17:41:09.860488     248 status_manager.go:631] "Status for pod updated successfully" pod="pods-7829/pod-submit-status-0-13" statusVersion=4 status={Phase:Pending Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-10-05 17:41:03 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-10-05 17:41:03 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [busybox]} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-10-05 17:41:03 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [busybox]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2021-10-05 17:41:03 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.2 PodIP:10.244.1.111 PodIPs:[{IP:10.244.1.111}] StartTime:2021-10-05 17:41:03 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:busybox State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:137,Signal:0,Reason:ContainerStatusUnknown,Message:The container could not be located when the pod was terminated,StartedAt:0001-01-01 00:00:00 +0000 UTC,FinishedAt:0001-01-01 00:00:00 +0000 UTC,ContainerID:,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:k8s.gcr.io/e2e-test-images/busybox:1.29-2 ImageID: ContainerID: Started:0xc0014d25a8}] QOSClass:Burstable EphemeralContainerStatuses:[]}
Oct 05 17:41:09 kind-worker kubelet[248]: I1005 17:41:09.860884     248 kubelet_pods.go:951] "Pod is terminated and all resources are reclaimed" pod="pods-7829/pod-submit-status-0-13"
Oct 05 17:41:09 kind-worker kubelet[248]: I1005 17:41:09.861840     248 kubelet.go:2068] "SyncLoop RECONCILE" source="api" pods=[pods-7829/pod-submit-status-0-13]
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.042381     248 request.go:597] Waited for 181.389859ms due to client-side throttling, not priority and fairness, request: DELETE:https://kind-control-plane:6443/api/v1/namespaces/pods-7829/pods/pod-submit-status-0-13
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.102100     248 kubelet.go:2071] "SyncLoop DELETE" source="api" pods=[pods-7829/pod-submit-status-0-13]
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.102555     248 pod_workers.go:557] "Pod is finished processing, no further updates" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.116535     248 status_manager.go:650] "Pod fully terminated and removed from etcd" pod="pods-7829/pod-submit-status-0-13"
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.119678     248 kubelet.go:2065] "SyncLoop REMOVE" source="api" pods=[pods-7829/pod-submit-status-0-13]
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.119730     248 kubelet.go:1910] "Pod has been deleted and must be killed" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7
Oct 05 17:41:10 kind-worker kubelet[248]: I1005 17:41:10.119758     248 pod_workers.go:557] "Pod is finished processing, no further updates" pod="pods-7829/pod-submit-status-0-13" podUID=a409f0f6-ac57-4d5e-95e1-97e074ea74b7
```

This just seems like accidental bad timing to me and unrelated to the PR... thoughts?

ehashman commented on test/e2e/node/pods.go:
```diff
@@ -425,7 +427,7 @@ var _ = SIGDescribe("Pods Extended", func() {
 								var names []string
 								for _, status := range pod.Status.ContainerStatuses {
-									if status.State.Terminated != nil || status.State.Running != nil {
+									if status.State.Running != nil {
```
I think this was just wrong before, but we never hit it because we were failing to sync out terminated container statuses. Now that we're doing that correctly, a Pending pod with a terminated container (perhaps because it was deleted before it reached Running) would cause this test to fail.

ehashman replied:
Note: this line came from https://github.com/kubernetes/kubernetes/pull/88440 which explicitly says in the description "The kubelet will consider any container without the `waiting` or `terminated` state to be still running, like it does for other operations." which doesn't match the code here.

ehashman commented:
/test pull-kubernetes-node-kubelet-eviction
/test pull-kubernetes-node-kubelet-serial

ehashman commented:
Passed tests :partying_face: 

The serial tests will fail, just wanna take a quick look for signal.

smarterclayton commented:
Are there any other optional jobs we should run?

ehashman commented:
Don't think so.

smarterclayton commented:
/lgtm
/approve
